# **UNSW_NB15 Threat Detection**

### **- Preprocessing, EDA, and Modelling**

#### **Objective**
The objective of this project is to make a classification model that classify the datasets

#### **Dataset**
The dataset used in this notebook is the **UNSW_NB15** [dataset](https://www.kaggle.com/datasets/mrwellsdavid/unsw-nb15/data).

### **Data Cleaning & Preprocessing**

##### importing libraries

In [1]:
import numpy as np
import pandas as pd
from zipfile import ZipFile
import os

##### defining path

In [2]:
home = r'D:\ml.projects\files\1_UNSW_Datasets.zip'
school = r'C:\Users\ncc\Desktop\ml.projects\files\1_UNSW_Datasets.zip'

if os.path.exists(home):
    path = home 
elif os.path.exists(school):
    path = school 
else:
    raise FileNotFoundError("Zip File Not Found on System")

##### loading & inspecting data

In [3]:
def load_data():
    with ZipFile(path, "r") as f:
        f.extract("UNSW_NB15_training-set.csv")
        f.extract("UNSW_NB15_testing-set.csv")
        f.extract("UNSW-NB15_LIST_EVENTS.csv")
        f.extract("UNSW-NB15_1.csv")
        f.extract("UNSW-NB15_2.csv")
        f.extract("UNSW-NB15_3.csv")
        f.extract("UNSW-NB15_4.csv")
        f.extract("NUSW-NB15_features.csv")
    train = pd.read_csv("UNSW_NB15_training-set.csv")
    test = pd.read_csv("UNSW_NB15_testing-set.csv")
    list_events = pd.read_csv("UNSW-NB15_LIST_EVENTS.csv")
    NB15_1 = pd.read_csv("UNSW-NB15_1.csv")
    NB15_2 = pd.read_csv("UNSW-NB15_2.csv")
    NB15_3 = pd.read_csv("UNSW-NB15_3.csv")
    NB15_4 = pd.read_csv("UNSW-NB15_4.csv")
    NB15_features = pd.read_csv("NUSW-NB15_features.csv", encoding= "cp1252")

    return train, test, list_events, NB15_1, NB15_2, NB15_3, NB15_4, NB15_features

In [4]:
train, test, list_events, NB15_1, NB15_2, NB15_3, NB15_4, NB15_features = load_data()

C:\Users\ncc\AppData\Local\Temp\ipykernel_9132\3277937790.py:14: DtypeWarning: Columns (1,3,47) have mixed types. Specify dtype option on import or set low_memory=False.
  NB15_1 = pd.read_csv("UNSW-NB15_1.csv")
C:\Users\ncc\AppData\Local\Temp\ipykernel_9132\3277937790.py:15: DtypeWarning: Columns (3,39,47) have mixed types. Specify dtype option on import or set low_memory=False.
  NB15_2 = pd.read_csv("UNSW-NB15_2.csv")


In [5]:
train.head()

id       dur proto service state  spkts  dpkts  sbytes  dbytes  \
0   1  0.000011   udp       -   INT      2      0     496       0   
1   2  0.000008   udp       -   INT      2      0    1762       0   
2   3  0.000005   udp       -   INT      2      0    1068       0   
3   4  0.000006   udp       -   INT      2      0     900       0   
4   5  0.000010   udp       -   INT      2      0    2126       0   

          rate  ...  ct_dst_sport_ltm  ct_dst_src_ltm  is_ftp_login  \
0   90909.0902  ...                 1               2             0   
1  125000.0003  ...                 1               2             0   
2  200000.0051  ...                 1               3             0   
3  166666.6608  ...                 1               3             0   
4  100000.0025  ...                 1               3             0   

   ct_ftp_cmd  ct_flw_http_mthd  ct_src_ltm  ct_srv_dst  is_sm_ips_ports  \
0           0                 0           1           2                0   
1           0                 0           1           2                0   
2           0                 0           1           3                0   
3           0                 0           2           3                0   
4           0                 0           2           3                0   

   attack_cat  label  
0      Normal      0  
1      Normal      0  
2      Normal      0  
3      Normal      0  
4      Normal      0  

[5 rows x 45 columns]

In [6]:
test.head()

id       dur proto service state  spkts  dpkts  sbytes  dbytes       rate  \
0   1  0.121478   tcp       -   FIN      6      4     258     172  74.087490   
1   2  0.649902   tcp       -   FIN     14     38     734   42014  78.473372   
2   3  1.623129   tcp       -   FIN      8     16     364   13186  14.170161   
3   4  1.681642   tcp     ftp   FIN     12     12     628     770  13.677108   
4   5  0.449454   tcp       -   FIN     10      6     534     268  33.373826   

   ...  ct_dst_sport_ltm  ct_dst_src_ltm  is_ftp_login  ct_ftp_cmd  \
0  ...                 1               1             0           0   
1  ...                 1               2             0           0   
2  ...                 1               3             0           0   
3  ...                 1               3             1           1   
4  ...                 1              40             0           0   

   ct_flw_http_mthd  ct_src_ltm  ct_srv_dst  is_sm_ips_ports  attack_cat  \
0                 0           1           1                0      Normal   
1                 0           1           6                0      Normal   
2                 0           2           6                0      Normal   
3                 0           2           1                0      Normal   
4                 0           2          39                0      Normal   

   label  
0      0  
1      0  
2      0  
3      0  
4      0  

[5 rows x 45 columns]

In [7]:
list_events.head()

Attack category Attack subcategory  Number of events
0          normal                NaN           2218761
1        Fuzzers                 FTP               558
2        Fuzzers                HTTP              1497
3        Fuzzers                 RIP              3550
4        Fuzzers                 SMB              5245

In [8]:
NB15_1.head()

59.166.0.0   1390  149.171.126.6    53  udp  CON  0.001055  132  164  31  \
0  59.166.0.0  33661  149.171.126.9  1024  udp  CON  0.036133  528  304  31   
1  59.166.0.6   1464  149.171.126.7    53  udp  CON  0.001119  146  178  31   
2  59.166.0.5   3593  149.171.126.5    53  udp  CON  0.001209  132  164  31   
3  59.166.0.3  49664  149.171.126.0    53  udp  CON  0.001169  146  178  31   
4  59.166.0.0  32119  149.171.126.9   111  udp  CON  0.078339  568  312  31   

   ...  0.17   3  7  1  3.1  1.1  1.2  1.3  Unnamed: 47  0.18  
0  ...     0   2  4  2    3    1    1    2          NaN     0  
1  ...     0  12  8  1    2    2    1    1          NaN     0  
2  ...     0   6  9  1    1    1    1    1          NaN     0  
3  ...     0   7  9  1    1    1    1    1          NaN     0  
4  ...     0   2  4  2    3    1    1    2          NaN     0  

[5 rows x 49 columns]

In [9]:
NB15_2.head()

59.166.0.0   6055  149.171.126.5  54145  tcp  FIN  0.072974  4238  60788  \
0  59.166.0.0   7832  149.171.126.3   5607  tcp  FIN  0.144951  5174  91072   
1  59.166.0.8  11397  149.171.126.6     21  tcp  FIN  0.116107  2934   3742   
2  59.166.0.0   3804  149.171.126.3     53  udp  CON  0.000986   146    178   
3  59.166.0.8  14339  149.171.126.6  14724  tcp  FIN  0.038480  8928    320   
4  59.166.0.8  39094  149.171.126.3     53  udp  CON  0.001026   130    162   

   31  ...  0.6  13  13.1  6  7.1  1  1.1  2  Unnamed: 47  0.7  
0  31  ...    0  13    13  6    7  1    1  2          NaN    0  
1  31  ...    1   1     2  7    5  1    1  4          NaN    0  
2  31  ...    0  13    13  6    7  1    1  2          NaN    0  
3  31  ...    0   8    20  7    5  1    1  4          NaN    0  
4  31  ...    0   8    13  6    5  1    1  1          NaN    0  

[5 rows x 49 columns]

In [10]:
NB15_3.head()

59.166.0.1  18247  149.171.126.4   7662  tcp  FIN  0.119596   4550  \
0      59.166.0.3  54771  149.171.126.2  27709  tcp  FIN  0.650574   8928   
1      59.166.0.8  13289  149.171.126.9   5190  tcp  FIN  0.007980   2158   
2  149.171.126.18   1043   175.45.176.3     53  udp  INT  0.000005    264   
3  149.171.126.18   1043   175.45.176.3     53  udp  INT  0.000005    264   
4      59.166.0.3  10275  149.171.126.0     25  tcp  FIN  0.486578  37462   

   68342  31  ...      6   2 2.1   5   1  1.1  2.2  Unnamed: 47  0.4  
0    320  31  ...      3   5   2   4   1    1    4          NaN    0  
1   2464  31  ...      3   5   1   1   1    1    3          NaN    0  
2      0  60  ...     19  19  19  19  19   19   19          NaN    0  
3      0  60  ...     19  19  19  19  19   19   19          NaN    0  
4   3380  31  ...      3   2   3   4   2    1    2          NaN    0  

[5 rows x 49 columns]

In [11]:
NB15_4.head()

59.166.0.9   7045  149.171.126.7    25  tcp  FIN  0.201886  37552     3380  \
0  59.166.0.9   9685  149.171.126.2    80  tcp  FIN  5.864748  19410  1087890   
1  59.166.0.2   1421  149.171.126.4    53  udp  CON  0.001391    146      178   
2  59.166.0.2  21553  149.171.126.2    25  tcp  FIN  0.053948  37812     3380   
3  59.166.0.8  45212  149.171.126.4    53  udp  CON  0.000953    146      178   
4  59.166.0.0  59922  149.171.126.8  6881  tcp  FIN  8.633186  25056  1094788   

   31  ...     2  2.1  7  4  1  1.1  3  Unnamed: 47  0.4  
0  31  ...     3    1  4  4  1    1  1          NaN    0  
1  31  ...     3    5  2  7  1    1  4          NaN    0  
2  31  ...     1    1  4  7  1    1  3          NaN    0  
3  31  ...     2    5  2  1  1    1  2          NaN    0  
4  31  ...     9    7  2  3  2    1  6          NaN    0  

[5 rows x 49 columns]

##### understading features

In [12]:
NB15_features

No.              Name      Type   \
0     1             srcip    nominal   
1     2             sport    integer   
2     3             dstip    nominal   
3     4            dsport    integer   
4     5             proto    nominal   
5     6             state    nominal   
6     7               dur      Float   
7     8            sbytes    Integer   
8     9            dbytes    Integer   
9    10              sttl    Integer   
10   11              dttl    Integer   
11   12             sloss    Integer   
12   13             dloss    Integer   
13   14           service    nominal   
14   15             Sload      Float   
15   16             Dload      Float   
16   17             Spkts    integer   
17   18             Dpkts    integer   
18   19              swin    integer   
19   20              dwin    integer   
20   21             stcpb    integer   
21   22             dtcpb    integer   
22   23           smeansz    integer   
23   24           dmeansz    integer   
24   25       trans_depth    integer   
25   26       res_bdy_len    integer   
26   27              Sjit      Float   
27   28              Djit      Float   
28   29             Stime  Timestamp   
29   30             Ltime  Timestamp   
30   31           Sintpkt      Float   
31   32           Dintpkt      Float   
32   33            tcprtt      Float   
33   34            synack      Float   
34   35            ackdat      Float   
35   36   is_sm_ips_ports     Binary   
36   37      ct_state_ttl    Integer   
37   38  ct_flw_http_mthd    Integer   
38   39      is_ftp_login     Binary   
39   40        ct_ftp_cmd    integer   
40   41        ct_srv_src    integer   
41   42        ct_srv_dst    integer   
42   43        ct_dst_ltm    integer   
43   44       ct_src_ ltm    integer   
44   45  ct_src_dport_ltm    integer   
45   46  ct_dst_sport_ltm    integer   
46   47    ct_dst_src_ltm    integer   
47   48        attack_cat    nominal   
48   49             Label     binary   

                                          Description  
0                                   Source IP address  
1                                  Source port number  
2                              Destination IP address  
3                             Destination port number  
4                                Transaction protocol  
5   Indicates to the state and its dependent proto...  
6                               Record total duration  
7            Source to destination transaction bytes   
8             Destination to source transaction bytes  
9           Source to destination time to live value   
10           Destination to source time to live value  
11           Source packets retransmitted or dropped   
12       Destination packets retransmitted or dropped  
13  http, ftp, smtp, ssh, dns, ftp-data ,irc  and ...  
14                             Source bits per second  
15                        Destination bits per second  
16                Source to destination packet count   
17                 Destination to source packet count  
18              Source TCP window advertisement value  
19         Destination TCP window advertisement value  
20                    Source TCP base sequence number  
21               Destination TCP base sequence number  
22  Mean of the ?ow packet size transmitted by the...  
23  Mean of the ?ow packet size transmitted by the...  
24  Represents the pipelined depth into the connec...  
25  Actual uncompressed content size of the data t...  
26                               Source jitter (mSec)  
27                          Destination jitter (mSec)  
28                                  record start time  
29                                   record last time  
30             Source interpacket arrival time (mSec)  
31        Destination interpacket arrival time (mSec)  
32  TCP connection setup round-trip time, the sum ...  
33  TCP connection setup time, the time between th...  
34  TCP connection setup time, the time bet

##### concating the data

In [13]:
# this replace all the column headers of NB15_(n)
# with the list of names in the "Name" column of NB15 features
NB15_1.columns = NB15_features['Name']  
NB15_2.columns = NB15_features['Name']
NB15_3.columns = NB15_features['Name']
NB15_4.columns = NB15_features['Name']

In [14]:
# joining the data together
train_df = pd.concat([NB15_1, NB15_2, NB15_3, NB15_4], ignore_index = True)

In [15]:
# shuffling data
train_df = train_df.sample(frac = 1, random_state = 42).reset_index(drop = True)

In [16]:
train_df

Name              srcip  sport           dstip dsport proto state       dur  \
0          175.45.176.3  57672  149.171.126.15   3260   tcp   CON  0.285356   
1            59.166.0.8  38052   149.171.126.9   6881   tcp   FIN  0.314311   
2            59.166.0.0  42911   149.171.126.2  38558   udp   CON  0.301180   
3          175.45.176.1  47439  149.171.126.14     53   udp   INT  0.000009   
4            59.166.0.5  61544   149.171.126.6     53   udp   CON  0.001079   
...                 ...    ...             ...    ...   ...   ...       ...   
2540038      59.166.0.8  34415   149.171.126.0   5190   tcp   FIN  0.008119   
2540039      59.166.0.8  56352   149.171.126.2     53   udp   CON  0.001047   
2540040      59.166.0.9  25527   149.171.126.4   6881   tcp   FIN  0.013106   
2540041    175.45.176.1  47439  149.171.126.14     53   udp   INT  0.000003   
2540042  149.171.126.14   1043    175.45.176.1     53   udp   INT  0.000002   

Name     sbytes  dbytes  sttl  ...  ct_ftp_cmd  ct_srv_src  ct_srv_dst  \
0           986      86    62  ...                      10           8   
1          1540    1644    31  ...                      14           6   
2           536     304    31  ...           0           9          16   
3           114       0   254  ...                      33          33   
4           146     178    31  ...                       2           2   
...         ...     ...   ...  ...         ...         ...         ...   
2540038    1920    4312    31  ...           0           3          10   
2540039     130     162    31  ...                       1           1   
2540040    1540    1644    31  ...                      12          11   
2540041     114       0   254  ...                      37          37   
2540042     264       0    60  ...                      22          22   

Name    ct_dst_ltm  ct_src_ ltm  ct_src_dport_ltm  ct_dst_sport_ltm  \
0                4            5                 3                 1   
1                8           12                 5                 1   
2                7            6                 1                 1   
3               17           17                17                17   
4                5            4                 1                 1   
...            ...          ...               ...               ...   
2540038          4            1                 1                 1   
2540039          2            6                 1                 1   
2540040          8            9                 6                 1   
2540041         26           26                26                13   
2540042         22           22                22                11   

Name     ct_dst_src_ltm  attack_cat  Label  
0                     7         NaN      0  
1                     6         NaN      0  
2                     3         NaN      0  
3                    33     Generic      1  
4                     1         NaN      0  
...                 ...         ...    ...  
2540038               1         NaN      0  
2540039               1         NaN      0  
2540040               6         NaN      0  
2540041              37     Generic      1  
2540042              22         NaN      0  

[2540043 rows x 49 columns]

#### **Analysis Data and Preprocessing**

In [17]:
def analysis(data):
    print(f"Shape of data is: {data.shape}\n")
    print(f"Statistical Summary")
    print('-------------------')
    print(data.info())

In [18]:
analysis(train_df)

Shape of data is: (2540043, 49)

Statistical Summary
-------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2540043 entries, 0 to 2540042
Data columns (total 49 columns):
 #   Column            Dtype  
---  ------            -----  
 0   srcip             object 
 1   sport             object 
 2   dstip             object 
 3   dsport            object 
 4   proto             object 
 5   state             object 
 6   dur               float64
 7   sbytes            int64  
 8   dbytes            int64  
 9   sttl              int64  
 10  dttl              int64  
 11  sloss             int64  
 12  dloss             int64  
 13  service           object 
 14  Sload             float64
 15  Dload             float64
 16  Spkts             int64  
 17  Dpkts             int64  
 18  swin              int64  
 19  dwin              int64  
 20  stcpb             int64  
 21  dtcpb             int64  
 22  smeansz           int64  
 23  dmeansz           int64  
 24  tra

**Insight:**

-   There are **2540043 records** and **49 columns** available in the train dataset.

In [19]:
cols = train_df.columns

numerical_cols = train_df.select_dtypes(include = ["int64", "float64"]).columns
categorical_cols = train_df.select_dtypes(include = ["object"]).columns

In [20]:
print(f"There are {len(numerical_cols)} Numerical Columns")
print(f"There are {len(categorical_cols)} Categorical Columns")

There are 40 Numerical Columns
There are 9 Categorical Columns


##### checking for duplicate values in the data

In [21]:
def duplicates(data):
    duplicated = data.duplicated().sum()
    print(f"There are {duplicated} duplicated values in the data")

    if duplicated.sum() > 0:
        print("Removing duplicates....")
        data = data.drop_duplicates()
        print("Duplicates removed!")
        print(f"There are now {data.duplicated().sum()} duplicates in the data")

In [22]:
duplicates(train_df)

There are 480626 duplicated values in the data
Removing duplicates....
Duplicates removed!
There are now 0 duplicates in the data


**Insight:**
-   There are **480626** duplicate records in the dataset.


##### computing total no. of missing values and the percentage

In [23]:
def missing_data(df):
    missing = df.isnull().sum().to_frame().rename(columns = {0: "Total No. of Missing Values"})
    missing['%of Missing Values'] = round((missing['Total No. of Missing Values']/len(df))*100, 2)
    return missing 

In [24]:
missing_data(train_df)

Total No. of Missing Values  %of Missing Values
Name                                                             
srcip                                       0                0.00
sport                                       0                0.00
dstip                                       0                0.00
dsport                                      0                0.00
proto                                       0                0.00
state                                       0                0.00
dur                                         0                0.00
sbytes                                      0                0.00
dbytes                                      0                0.00
sttl                                        0                0.00
dttl                                        0                0.00
sloss                                       0                0.00
dloss                                       0                0.00
service                                     0                0.00
Sload                                       0                0.00
Dload                                       0                0.00
Spkts                                       0                0.00
Dpkts                                       0                0.00
swin                                        0                0.00
dwin                                        0                0.00
stcpb                                       0                0.00
dtcpb                                       0                0.00
smeansz                                     0                0.00
dmeansz                                     0                0.00
trans_depth                                 0                0.00
res_bdy_len                                 0                0.00
Sjit                                        0                0.00
Djit                                        0                0.00
Stime                                       0                0.00
Ltime                                       0                0.00
Sintpkt                                     0                0.00
Dintpkt                                     0                0.00
tcprtt                                      0                0.00
synack                                      0                0.00
ackdat                                      0                0.00
is_sm_ips_ports                             0                0.00
ct_state_ttl                                0                0.00
ct_flw_http_mthd                      1348143               53.08
is_ftp_login                          1429877               56.29
ct_ftp_cmd                                  0                0.00
ct_srv_src                                  0                0.00
ct_srv_dst                                  0                0.00
ct_dst_ltm                                  0                0.00
ct_src_ ltm                                 0                0.00
ct_src_dport_ltm                            0                0.00
ct_dst_sport_ltm                            0                0.00
ct_dst_src_ltm                              0                0.00
attack_cat                            2218760               87.35
Label                                       0                0.00

##### handling missing values

In [25]:
def handling_missing_values(df):
    # replacing null values in 'attack_cat' with 'normal'
    df['attack_cat'].fillna('Normal', inplace = True)
    df['attack_cat'] = df['attack_cat'].apply(lambda x: x.strip().lower())

    # replacing null values in 'is_ftp_login' with 0
    df['is_ftp_login'].fillna(0, inplace = True)

    # replacing null values in 'ct_flw_http_mthd' with 0
    df['ct_flw_http_mthd'].fillna(0, inplace = True)
    return df 

In [26]:
train_df = handling_missing_values(train_df)

C:\Users\ncc\AppData\Local\Temp\ipykernel_9132\1601984324.py:3: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['attack_cat'].fillna('Normal', inplace = True)
C:\Users\ncc\AppData\Local\Temp\ipykernel_9132\1601984324.py:7: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example,

**Insight:**
-   `attack_cat`, `is_ftp_login`, and `ct_flw_http_mthd` all have missing values.

#### **Performing Descriptive Analysis**

In [27]:
train_df.describe().T

count          mean           std           min  \
Name                                                                    
dur               2540043.0  6.587925e-01  1.392494e+01  0.000000e+00   
sbytes            2540043.0  4.339589e+03  5.640604e+04  0.000000e+00   
dbytes            2540043.0  3.642760e+04  1.610962e+05  0.000000e+00   
sttl              2540043.0  6.278202e+01  7.462281e+01  0.000000e+00   
dttl              2540043.0  3.076681e+01  4.285092e+01  0.000000e+00   
sloss             2540043.0  5.163917e+00  2.251709e+01  0.000000e+00   
dloss             2540043.0  1.632944e+01  5.659479e+01  0.000000e+00   
Sload             2540043.0  3.695650e+07  1.186044e+08  0.000000e+00   
Dload             2540043.0  2.450860e+06  4.224865e+06  0.000000e+00   
Spkts             2540043.0  3.328882e+01  7.628393e+01  0.000000e+00   
Dpkts             2540043.0  4.272663e+01  1.215021e+02  0.000000e+00   
swin              2540043.0  1.500887e+02  1.254824e+02  0.000000e+00   
dwin              2540043.0  1.497458e+02  1.255439e+02  0.000000e+00   
stcpb             2540043.0  1.261702e+09  1.422028e+09  0.000000e+00   
dtcpb             2540043.0  1.261766e+09  1.422139e+09  0.000000e+00   
smeansz           2540043.0  1.242535e+02  1.519158e+02  0.000000e+00   
dmeansz           2540043.0  2.766716e+02  3.356165e+02  0.000000e+00   
trans_depth       2540043.0  8.325332e-02  3.500048e-01  0.000000e+00   
res_bdy_len       2540043.0  4.242124e+03  4.750056e+04  0.000000e+00   
Sjit              2540043.0  1.589039e+03  1.691038e+04  0.000000e+00   
Djit              2540043.0  7.300766e+02  3.438561e+03  0.000000e+00   
Stime             2540043.0  1.423261e+09  1.134448e+06  1.421927e+09   
Ltime             2540043.0  1.423261e+09  1.134448e+06  1.421927e+09   
Sintpkt           2540043.0  1.933228e+02  2.779165e+03  0.000000e+00   
Dintpkt           2540043.0  7.882489e+01  1.433193e+03  0.000000e+00   
tcprtt            2540043.0  6.180483e-03  4.615867e-02  0.000000e+00   
synack            2540043.0  3.287599e-03  2.593572e-02  0.000000e+00   
ackdat            2540043.0  2.892884e-03  2.394670e-02  0.000000e+00   
is_sm_ips_ports   2540043.0  1.651547e-03  4.060566e-02  0.000000e+00   
ct_state_ttl      2540043.0  2.611550e-01  6.830978e-01  0.000000e+00   
ct_flw_http_mthd  2540043.0  1.100781e-01  5.564199e-01  0.000000e+00   
is_ftp_login      2540043.0  1.735128e-02  1.334573e-01  0.000000e+00   
ct_srv_src        2540043.0  9.206993e+00  1.083676e+01  1.000000e+00   
ct_srv_dst        2540043.0  8.988962e+00  1.082250e+01  1.000000e+00   
ct_dst_ltm        2540043.0  6.439107e+00  8.162039e+00  1.000000e+00   
ct_src_ ltm       2540043.0  6.900989e+00  8.205068e+00  1.000000e+00   
ct_src_dport_ltm  2540043.0  4.642145e+00  8.477584e+00  1.000000e+00   
ct_dst_sport_ltm  2540043.0  3.592733e+00  6.174449e+00  1.000000e+00   
ct_dst_src_ltm    2540043.0  6.845894e+00  1.125829e+01  1.000000e+00   
Label             2540043.0  1.264872e-01  3.323977e-01  0.000000e+00   

                           25%           50%           75%           max  
Name                                                                      
dur               1.037000e-03  1.586100e-02  2.145650e-01  8.786638e+03  
sbytes            2.000000e+02  1.470000e+03  3.182000e+03  1.435577e+07  
dbytes            1.780000e+02  1.820000e+03  1.489400e+04  1.465753e+07  
sttl              3.100000e+01  3.100000e+01  3.100000e+01  2.550000e+02  
dttl              2.900000e+01  2.900000e+01  2.900000e+01  2.540000e+02  
sloss             0.000000e+00  3.000000e+00  7.000000e+00  5.319000e+03  
dloss             0.000000e+00  4.000000e+00  1.400000e+01  5.507000e+03  
Sload             1.353955e+05  5.893038e+05  2.039923e+06  5.988000e+09  
Dload             1.191580e+04  5.893179e+05  2.925945e+06  1.287619e+08  
Spkts             2.000000e+00  1.200000e+01  4.400000e+01  1.064600e+04  
Dpkts             2.000000e+00  1.200000e+01  4.200000e

#### **Performing Descriptive Analysis on Categorical Attributes**

In [28]:
train_df.describe(include = "O").T

count  unique            top     freq
Name                                               
srcip       2540043      43     59.166.0.4   197959
sport       2540043  100343           1043   216289
dstip       2540043      47  149.171.126.3   197771
dsport      2540043  128297             53   619539
proto       2540043     135            tcp  1495071
state       2540043      16            FIN  1478686
service     2540043      13              -  1246395
ct_ftp_cmd  2540043      13                 1429877
attack_cat  2540043      11         normal  2218760

#### **Checking Unique Values of Attributes**

In [29]:
for column in cols:
    print(f"Unique Values of {column} with dtype {train_df[column].dtype} have total values - {train_df[column].nunique()} -> {train_df[column].unique()}")
    print("*"*100)

Unique Values of srcip with dtype object have total values - 43 -> ['175.45.176.3' '59.166.0.8' '59.166.0.0' '175.45.176.1' '59.166.0.5'
 '59.166.0.7' '175.45.176.0' '149.171.126.14' '59.166.0.1' '59.166.0.9'
 '149.171.126.15' '149.171.126.18' '59.166.0.4' '59.166.0.2' '59.166.0.3'
 '59.166.0.6' '175.45.176.2' '149.171.126.10' '10.40.182.1'
 '149.171.126.12' '10.40.170.2' '10.40.85.30' '149.171.126.2'
 '10.40.182.3' '10.40.85.1' '10.40.182.6' '10.40.85.10' '149.171.126.7'
 '149.171.126.6' '149.171.126.0' '149.171.126.8' '149.171.126.13'
 '149.171.126.5' '192.168.241.243' '149.171.126.4' '149.171.126.9'
 '149.171.126.1' '149.171.126.3' '149.171.126.11' '149.171.126.19'
 '149.171.126.17' '149.171.126.16' '127.0.0.1']
****************************************************************************************************
Unique Values of sport with dtype object have total values - 100343 -> [57672 38052 42911 ... '17953' '20349' '53235']
*******************************************************

**Insight:**
-   Some values were repeated due to their wrong format, e.g. column `ct_ftp_cmd` unique values are `[0'''1' 1 '0' 2 8 '2' 4 5 3 6 '4']` in which we see that same values repeated and treated differently.
-   In column `is_ftp_login`, the unique values are `[0. 1. 2. 4.]` which is not correct because according to given NB15_features.csv column is a binary column

**Handling ct_ftp_cmd**

In [ ]:
def ct_ftp_cmd(df, column):
    df[column] = df[column].astype('str')
    df[column].dtype

    df[column].unique()
    df[column] = df[column].replace(' ', '0')
    df[column] = df[column].astype('int')
    return df

In [ ]:
ct_ftp_cmd(train_df, 'ct_ftp_cmd')

Name              srcip  sport           dstip dsport proto state       dur  \
0          175.45.176.3  57672  149.171.126.15   3260   tcp   CON  0.285356   
1            59.166.0.8  38052   149.171.126.9   6881   tcp   FIN  0.314311   
2            59.166.0.0  42911   149.171.126.2  38558   udp   CON  0.301180   
3          175.45.176.1  47439  149.171.126.14     53   udp   INT  0.000009   
4            59.166.0.5  61544   149.171.126.6     53   udp   CON  0.001079   
...                 ...    ...             ...    ...   ...   ...       ...   
2540038      59.166.0.8  34415   149.171.126.0   5190   tcp   FIN  0.008119   
2540039      59.166.0.8  56352   149.171.126.2     53   udp   CON  0.001047   
2540040      59.166.0.9  25527   149.171.126.4   6881   tcp   FIN  0.013106   
2540041    175.45.176.1  47439  149.171.126.14     53   udp   INT  0.000003   
2540042  149.171.126.14   1043    175.45.176.1     53   udp   INT  0.000002   

Name     sbytes  dbytes  sttl  ...  ct_ftp_cmd  ct_srv_src  ct_srv_dst  \
0           986      86    62  ...           0          10           8   
1          1540    1644    31  ...           0          14           6   
2           536     304    31  ...           0           9          16   
3           114       0   254  ...           0          33          33   
4           146     178    31  ...           0           2           2   
...         ...     ...   ...  ...         ...         ...         ...   
2540038    1920    4312    31  ...           0           3          10   
2540039     130     162    31  ...           0           1           1   
2540040    1540    1644    31  ...           0          12          11   
2540041     114       0   254  ...           0          37          37   
2540042     264       0    60  ...           0          22          22   

Name    ct_dst_ltm  ct_src_ ltm  ct_src_dport_ltm  ct_dst_sport_ltm  \
0                4            5                 3                 1   
1                8           12                 5                 1   
2                7            6                 1                 1   
3               17           17                17                17   
4                5            4                 1                 1   
...            ...          ...               ...               ...   
2540038          4            1                 1                 1   
2540039          2            6                 1                 1   
2540040          8            9                 6                 1   
2540041         26           26                26                13   
2540042         22           22                22                11   

Name     ct_dst_src_ltm  attack_cat  Label  
0                     7      normal      0  
1                     6      normal      0  
2                     3      normal      0  
3                    33     generic      1  
4                     1      normal      0  
...                 ...         ...    ...  
2540038               1      normal      0  
2540039               1      normal      0  
2540040               6      normal      0  
2540041              37     generic      1  
2540042              22      normal      0  

[2540043 rows x 49 columns]

**Handling is_ftp_login**

In [48]:
def con_to_bin(df, column):
    df[column] = (df[column] > 0).astype(int)
    return df

In [49]:
con_to_bin(train_df, 'is_ftp_login')

Name              srcip  sport           dstip dsport proto state       dur  \
0          175.45.176.3  57672  149.171.126.15   3260   tcp   CON  0.285356   
1            59.166.0.8  38052   149.171.126.9   6881   tcp   FIN  0.314311   
2            59.166.0.0  42911   149.171.126.2  38558   udp   CON  0.301180   
3          175.45.176.1  47439  149.171.126.14     53   udp   INT  0.000009   
4            59.166.0.5  61544   149.171.126.6     53   udp   CON  0.001079   
...                 ...    ...             ...    ...   ...   ...       ...   
2540038      59.166.0.8  34415   149.171.126.0   5190   tcp   FIN  0.008119   
2540039      59.166.0.8  56352   149.171.126.2     53   udp   CON  0.001047   
2540040      59.166.0.9  25527   149.171.126.4   6881   tcp   FIN  0.013106   
2540041    175.45.176.1  47439  149.171.126.14     53   udp   INT  0.000003   
2540042  149.171.126.14   1043    175.45.176.1     53   udp   INT  0.000002   

Name     sbytes  dbytes  sttl  ...  ct_ftp_cmd  ct_srv_src  ct_srv_dst  \
0           986      86    62  ...           0          10           8   
1          1540    1644    31  ...           0          14           6   
2           536     304    31  ...           0           9          16   
3           114       0   254  ...           0          33          33   
4           146     178    31  ...           0           2           2   
...         ...     ...   ...  ...         ...         ...         ...   
2540038    1920    4312    31  ...           0           3          10   
2540039     130     162    31  ...           0           1           1   
2540040    1540    1644    31  ...           0          12          11   
2540041     114       0   254  ...           0          37          37   
2540042     264       0    60  ...           0          22          22   

Name    ct_dst_ltm  ct_src_ ltm  ct_src_dport_ltm  ct_dst_sport_ltm  \
0                4            5                 3                 1   
1                8           12                 5                 1   
2                7            6                 1                 1   
3               17           17                17                17   
4                5            4                 1                 1   
...            ...          ...               ...               ...   
2540038          4            1                 1                 1   
2540039          2            6                 1                 1   
2540040          8            9                 6                 1   
2540041         26           26                26                13   
2540042         22           22                22                11   

Name     ct_dst_src_ltm  attack_cat  Label  
0                     7      normal      0  
1                     6      normal      0  
2                     3      normal      0  
3                    33     generic      1  
4                     1      normal      0  
...                 ...         ...    ...  
2540038               1      normal      0  
2540039               1      normal      0  
2540040               6      normal      0  
2540041              37     generic      1  
2540042              22      normal      0  

[2540043 rows x 49 columns]

In [ ]:
def ports(df, column):
    # convert to string first
    df[column] = df[column].astype('str')

    # then handling non-integer values
    df[column] = pd.to_numeric(df[column], errors = 'coerce').fillna(0).astype(int)

In [54]:
ports(train_df, 'sport')

In [56]:
ports(train_df, 'dsport')

## Exploratory Data Analysis